src/main/java/com/google/api/codegen/transformer/go/GoModelTypeNameConverter.java

```diff
@@ -0,0 +1,193 @@
+          .put(Type.TYPE_SFIXED64, "int64")
+          .put(Type.TYPE_INT32, "int32")
+          .put(Type.TYPE_UINT32, "uint32")
+          .put(Type.TYPE_SINT32, "int32")
+          .put(Type.TYPE_FIXED32, "int32")
+          .put(Type.TYPE_SFIXED32, "int32")
+          .put(Type.TYPE_STRING, "string")
+          .put(Type.TYPE_BYTES, "[]byte")
+          .build();
+
+  @Override
+  public TypeName getTypeName(TypeRef type) {
+    if (type.isMap()) {
+      TypeName keyTypeName = getTypeNameForElementType(type.getMapKeyField().getType());
+      TypeName valueTypeName = getTypeNameForElementType(type.getMapValueField().getType());
+      return new TypeName("", "", "map[%i]%i", keyTypeName, valueTypeName);
+    } else if (type.isRepeated()) {
+      TypeName elementTypeName = getTypeNameForElementType(type);
+      return new TypeName("", "", "[]%i", elementTypeName);
+    } else {
+      return getTypeNameForElementType(type);
+    }
+  }
+
+  @Override
+  public TypeName getTypeNameForElementType(TypeRef type) {
+    String primitiveTypeName = PRIMITIVE_TYPE_MAP.get(type.getKind());
+    if (primitiveTypeName != null) {
+      return new TypeName(primitiveTypeName);
+    }
+    switch (type.getKind()) {
+      case TYPE_MESSAGE:
+        return getTypeName(type.getMessageType());
+      case TYPE_ENUM:
+        return getTypeName(type.getEnumType());
+      default:
+        throw new IllegalArgumentException("unknown type kind: " + type.getKind());
+    }
+  }
+
+  /**
+   * Since Go imports by package instead of by class name,
+   * we have to treat it differently than other languages.
+   * Imports in Go have 3 components:
+   * - Import path, like "github.com/googleapis/gax-go"
+   * - local package name, like "gax"; this is the name we call the package we import
+   * - The name of the thing in the package we want, like "CallOption"
+   *
+   * We need all 3 pieces in the full name and the last 2 in the nickname.
+   * For nickname, we simply use: "gax.CallOption",
+   * which is how we refer to the import in the program text.
+   * For the full name, we use join all 3 by semicolons:
+   * "github.com/googleapis/gax-go;gax;CallOption"
+   */
+  @Override
+  public TypeName getTypeName(ProtoElement elem) {
+    String importPath = elem.getFile().getProto().getOptions().getGoPackage();
+    String protoPackage = elem.getFile().getProto().getPackage();
+    String elemName = elem.getSimpleName();
+
+    // This is out of our list of curated protos,
+    // and we don't currently have a good way to name them.
+    if (!Strings.isNullOrEmpty(importPath) && !importPath.startsWith(CORE_PROTO_BASE)) {
+      String localName = protoPackage.replace(".", "_");
+      return new TypeName(
+          Joiner.on(";").join(importPath, localName, elemName), localName + "." + elemName);
+    }
+
+    // If the go_package option doesn't exist, infer it.
+    if (Strings.isNullOrEmpty(importPath)) {
+      String GOOGLE_PREFIX = "google.";
```

Comment:
The name is a lie - in that for it to be upper underscore, it has to be a constant (final). Maybe move this into the class at large instead of being a local.